experimental/components/src/units/converter_factory.rs

```diff
@@ -0,0 +1,339 @@
+    payload: &'data UnitsInfoV1<'data>,
+    payload_store: &'data ZeroTrieSimpleAscii<ZeroVec<'data, u8>>,
+}
+
+impl From<Sign> for num_bigint::Sign {
+    fn from(val: Sign) -> Self {
+        match val {
+            Sign::Positive => num_bigint::Sign::Plus,
+            Sign::Negative => num_bigint::Sign::Minus,
+        }
+    }
+}
+
+impl<'data> ConverterFactory<'data> {
+    #[cfg(feature = "datagen")]
+    pub fn from_payload(
+        payload: &'data UnitsInfoV1<'data>,
+        payload_store: &'data ZeroTrieSimpleAscii<ZeroVec<'data, u8>>,
+    ) -> Self {
+        Self {
+            payload,
+            payload_store,
+        }
+    }
+
+    pub fn parser(&self) -> MeasureUnitParser<'_> {
+        MeasureUnitParser::from_payload(self.payload_store)
+    }
+
+    /// Computes the offset between the input and output units.
+    /// To calculate the offset, we follow these steps:
+    /// 1. Assume the conversion rate for unit 1 to the root is: N1/D1 + OffsetN1/OffsetD1.
+    /// 2. Assume the conversion rate for unit 2 to the root is: N2/D2 + OffsetN2/OffsetD2.
+    /// 3. To convert from the root to unit 2, the formula is: D2/N2 - OffsetN2/OffsetD2 * (D2/N2).
+    /// 4. Given a value V to be converted from unit 1 to unit 2, the conversion will be the root:
+    ///    (V * N1/D1) + OffsetN1/OffsetD1, denoted as V_Root.
+    /// 5. To convert V_Root to unit 2, the formula is: V_Root * D2/N2 - OffsetN2/OffsetD2 * D2/N2.
+    /// 6. Substituting V_Root from step 4 into step 5 yields:
+    ///    ((V * N1/D1) + OffsetN1/OffsetD1) * D2/N2 - OffsetN2/OffsetD2 * (D2/N2).
+    /// 7. Simplifying the equation gives:
+    ///    V * (N1/D1) * (D2/N2) + OffsetN1/OffsetD1 * (D2/N2) - OffsetN2/OffsetD2 * (D2/N2).
+    /// 8. Focusing on the constants (offsets), we derive the offset formula:
+    ///    Offset = (OffsetN1/OffsetD1 - OffsetN2/OffsetD2) * (D2/N2).
+    ///    This simplifies to: Offset = (Offset1 - Offset2) * (1/ConversionRate of Unit2).
+    ///
+    /// NOTE:
+    ///   In order to have an offset, the input and output units should be simple.
+    /// This means, the input and output units should have only one unit item with power 1 and si prefix 0.
+    /// For example:
+    ///           1 - `meter` and `foot` are simple units.
+    ///           2 - `meter-per-second` and `foot-per-second` are not simple units.
+    fn compute_offset(
+        &self,
+        input_unit: &MeasureUnit,
+        output_unit: &MeasureUnit,
+    ) -> Option<Ratio<BigInt>> {
+        if !(input_unit.contained_units.len() == 1
+            && output_unit.contained_units.len() == 1
+            && input_unit.contained_units[0].power == 1
+            && output_unit.contained_units[0].power == 1
+            && input_unit.contained_units[0].si_prefix.power == 0
+            && output_unit.contained_units[0].si_prefix.power == 0)
+        {
+            return Some(Ratio::<BigInt>::from_integer(0.into()));
```

Comment:
Issue: The OffsetConverter should fail to build if the narrow conditions for offset calculations are not met. It shouldn't throw away the offset and return 0.

Further, given the narrow condition (only one unit on each side and no other factors involved), it should be possible to significantly simplify this function.

Reply:
I am returning `zero` because I want to differentiate between two cases:

1. There is an error in getting the offset which I return `Option::None`
2. The index is zero and in this case I return `ProportionalConverter`